File: elixir/plugins/httpjson.ex
# The round-trip every remote store shares - and OUTSIDE THE CORE with the
# client under it.
#
# Eight of the ten plugin kinds read a vault over HTTPS, and all eight
# want the same three things: one JSON round-trip whose failures are
# never a miss, a login token renewed shortly before its lease runs out,
# and a handful of conversions between a JSON value and the string a
# secret is. Written once here rather than eight times, and never in the
# core: a chain of the four built-ins must not link an HTTP client.
#
# `runcmd` is NOT here. The two kinds that read a store through its own
# CLI - boru and secretspec - reach `Sekreto.Plugins.Proc` instead, so
# that secretspec, which opens no socket at all, loads no socket code.

defmodule Sekreto.Plugins.Httpjson do
  @moduledoc """
  The shared plugin edge: one JSON round-trip, token renewal, and the
  small conversions the vault clients share.

  A plugin module: nothing under `src/` names it. See
  docs/design/plugin-providers.md.
  """

  alias Sekreto.Cell
  alias Sekreto.Error
  alias Sekreto.Json
  alias Sekreto.Plugins.Http

  @doc """
  The renewal time of a token that has no expiry: never.

  A configured token is kept forever; only a logged-in one carries a
  lease. Answered by a function rather than held in a module attribute,
  because every plugin that keeps a token cell needs the same value and
  an attribute cannot cross a module.
  """
  def never, do: 9_223_372_036_854_775_807

  @doc "An address without its trailing slash."
  def trimslash(text), do: Sekreto.dropsuffix(text, "/")

  @doc """
  One JSON round-trip. Answers `%{status: integer, body: json | :none}`.

  Network failure is always an error - an unreachable store is a store that
  could not answer. A success status promised JSON, so a 200 whose body
  does not parse is a store that answered incoherently, and treating that
  as a miss would fall through to a weaker store. Error statuses may carry
  any body at all: they are decided on status alone.
  """
  def fetchjson(method, url, headers \\ [], body \\ nil) do
    answer = Http.fetch(method, url, headers, body)

    parsed =
      case Json.parse(answer.body) do
        {:ok, value} -> value
        :error -> :none
      end

    if 200 == answer.status and :none == parsed do
      raise Error, message: "sekreto: malformed response from " <> Http.bare(url)
    end

    %{status: answer.status, body: parsed}
  end

  @doc """
  When a logged-in token must be renewed, from its expiry in seconds (a
  JSON number, or a string as Azure IMDS sends it): now + max(seconds - 60,
  1). A missing or zero expiry means never renew.
  """
  def renewtime(expires) do
    seconds =
      case expires do
        {:num, value} ->
          value

        {:str, value} ->
          case Float.parse(value) do
            {value, _rest} -> value
            :error -> 0.0
          end

        _other ->
          0.0
      end

    if 0 >= seconds do
      never()
    else
      System.system_time(:millisecond) + trunc(max(seconds - 60, 1.0) * 1000)
    end
  end

  @doc """
  Is this token cell due for renewal? A cell holding no token has never
  logged in; one past its renewal time is about to expire.
  """
  def expired?(cell) do
    held = Cell.get(cell)
    "" == held.token or System.system_time(:millisecond) >= held.renewat
  end

  @doc """
  Strict base64.

  Whitespace is stripped first - the canonical function accepts embedded
  newlines - and then anything outside the standard alphabet, or a length
  that is not a multiple of four, is REFUSED. A lenient decoder silently
  skips what it does not recognise and hands back plausible bytes for a
  corrupted payload, which then get returned as the secret.
  """
  def unbase64(text) do
    stripped = String.replace(text, ~r/\s/, "")

    case Base.decode64(stripped, padding: true) do
      {:ok, bytes} -> bytes
      :error -> :error
    end
  end

  @doc "A JSON accessor's `:none` as the miss the chain reads: nil."
  def nonone(:none), do: nil
  def nonone(value), do: value

  @doc "Any value as text, for a message: a status code, a kv version."
  def tostr(value) when is_binary(value), do: value
  def tostr(value) when is_integer(value), do: Integer.to_string(value)
  def tostr(value) when is_float(value), do: Json.numstr(value)
  def tostr(value), do: inspect(value)
end

File: ocaml/plugins/httpjson.ml
(* One JSON round-trip, and the reads a response body needs.

   THE SHARED HTTP-JSON HELPER, AND IT LIVES UNDER plugins/ BECAUSE OF
   THAT. Seven of the ten plugin kinds dial a store over HTTPS; the four
   built-in kinds dial nothing, so a chain of built-ins must never link
   this module, the framing under it or the TLS binding under that. *)

open Secret

(* A url without its query string, for a message that must not leak one. *)
let bare = Http.bare

(* One JSON round-trip's result: the status, and the parsed body. *)
type answer = { status : int; jbody : Json.t option }

(* One JSON round-trip. Network failure is always an error - an unreachable
   store is a store that could not answer, never a store without the
   secret. *)
let fetchjson ?(headers = []) ?body (meth : string) (url : string) : answer =
  let res = Http.request meth url headers body in
  let parsed = Json.parse res.Http.body in

  (* A success status promised JSON; a body that does not parse means the
     store could not answer coherently, and treating it as a miss would fall
     through to a weaker store. Error statuses may carry any body - they are
     decided on status alone. *)
  if 200 = res.Http.status && None = parsed then
    fail ("sekreto: malformed response from " ^ bare url);

  { status = res.Http.status; jbody = parsed }

(* When a logged-in token must be renewed, from its expiry in seconds (a
   JSON number, or a string as Azure IMDS sends it): now + max(seconds - 60,
   1). A missing or zero expiry means never renew. *)
let never = infinity

let renewtime (expires : Json.t option) : float =
  let seconds =
    match expires with
    | Some (Json.Num value) -> value
    | Some (Json.Str value) -> ( match float_of_string_opt value with Some v -> v | None -> 0.0)
    | _ -> 0.0
  in
  if Float.is_nan seconds || 0.0 >= seconds then never
  else (Unix.gettimeofday () *. 1000.0) +. (Float.max (seconds -. 60.0) 1.0 *. 1000.0)

let nowms () : float = Unix.gettimeofday () *. 1000.0

(* ---- reading a response ---------------------------------------------- *)

(* The text at a path in a response body, or nothing. A JSON null is nothing
   - a null field is a miss, never the string `null`. *)
let digtext (body : Json.t option) (keys : string list) : string option =
  Json.otext (Json.odig body keys)

(* The string at a path, insisting it really is a string: `__type` must be
   one, a 1Password vault list must be an array, a Doppler config an
   object. *)
let digstr (body : Json.t option) (keys : string list) : string option =
  Json.oasstr (Json.odig body keys)

let isset (value : string option) : bool =
  match value with Some text -> "" <> text | None -> false
